pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,416 @@
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HEADObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const DirDelim = "/"
+
+// Bucket implements the store.Bucket interface against cos-compatible(Tencent Object Storage) APIs.
+type Bucket struct {
+	logger   log.Logger
+	client   *cos.Client
+	opsTotal *prometheus.CounterVec
+}
+
+// Config encapsulates the necessary config values to instantiate an cos client.
+type Config struct {
+	Bucket    string
+	AppId     string
+	Region    string
+	SecretId  string
+	SecretKey string
+}
+
+// RegisterCosParams registers the cos flags and returns an initialized Config struct.
+func RegisterCosParams(cmd *kingpin.CmdClause) *Config {
+	var cosConfig Config
+
+	cmd.Flag("cos.bucket", "Cos-Compatible API bucket name for stored blocks.").
+		PlaceHolder("<bucket>").Envar("COS_BUCKET").StringVar(&cosConfig.Bucket)
+
+	cmd.Flag("cos.appid", "Cos-Compatible API endpoint for stored blocks.").
+		PlaceHolder("<appid>").Envar("COS_APPID").StringVar(&cosConfig.AppId)
+
+	cmd.Flag("cos.region", "Cos-Compatible API region for stored blocks.").
+		PlaceHolder("<region>").Envar("COS_REGION").StringVar(&cosConfig.Region)
+
+	cosConfig.SecretKey = os.Getenv("COS_SECRET_KEY")
+	cosConfig.SecretId = os.Getenv("COS_SECRET_ID")
+
+	return &cosConfig
+}
+
+// Validate checks to see if mandatory cos config options are set.
+func (conf *Config) Validate() error {
+	if conf.Bucket == "" ||
+		conf.AppId == "" ||
+		conf.Region == "" ||
+		conf.SecretId == "" ||
+		conf.SecretKey == "" {
+		return errors.New("insufficient cos configuration information")
+	}
+	return nil
+}
+
+func NewBucket(logger log.Logger, conf *Config, reg prometheus.Registerer, component string) (*Bucket, error) {
+	if nil == logger {
+		logger = log.NewNopLogger()
+	}
+
+	bucketUrl := cos.NewBucketURL(conf.Bucket, conf.AppId, conf.Region, true)
+
+	b, err := cos.NewBaseURL(bucketUrl.String())
+	if nil != err {
+		return nil, errors.Wrap(err, "initialize cos base url")
+	}
+
+	client := cos.NewClient(b, &http.Client{
+		Transport: &cos.AuthorizationTransport{
+			SecretID:  conf.SecretId,
+			SecretKey: conf.SecretKey,
+		},
+	})
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_cos_bucket_operations_total",
+			Help:        "Total number of operations that were executed against an cos bucket.",
+			ConstLabels: prometheus.Labels{"bucket": conf.Bucket},
+		}, []string{"operation"}),
+	}
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	b.opsTotal.WithLabelValues(opObjectInsert).Inc()
+
+	reader, length, err := objectLength(r)
```

Comment:
At first, it is no need to calculate the length of buffer, but the request header of COS should provide the length of body uploaded as `Content-Length`, it will not calculate the length of `*os.File` automatically, and it will be failed to upload the block file. I'm unhappy with sad resignation. It looks good when I do unit test. So maybe we should update the `objstore` unit test with mock block files as `*os.File` to find this problem.

Reply:
@bwplotka I confirmed that all methods invoked the method will pass the `io.Reader` as `*os.File` except for unit test. So should we update the `objstore` unit test with mock block files as `*os.File`?

Reply:
yes, please, that would remove the production code that do the buffering.